cmd/sts-handlers.go

```diff
@@ -411,3 +424,143 @@ func (sts *stsAPIHandlers) AssumeRoleWithWebIdentity(w http.ResponseWriter, r *h
+		return
+	}
+
+	action := r.Form.Get("Action")
+	switch action {
+	case ldapIdentity:
+	default:
+		logger.LogIf(ctx, fmt.Errorf("Unsupported action %s", action))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	ctx = newContext(r, w, action)
+	defer logger.AuditLog(w, r, action, nil)
+
+	ldapUsername := r.Form.Get("LDAPUsername")
+	ldapPassword := r.Form.Get("LDAPPassword")
+
+	if ldapUsername == "" || ldapPassword == "" {
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSMissingParameter))
+		return
+	}
+
+	ldapConn, err := globalServerConfig.LDAPServerConfig.Connect()
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("LDAP server connection failure: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+	if ldapConn == nil {
+		logger.LogIf(ctx, fmt.Errorf("LDAP server not configured: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	usernameSubs := newSubstituter("username", ldapUsername)
+	// We ignore error below as we already validated the username
+	// format string at startup.
+	usernameDN, _ := usernameSubs.substitute(globalServerConfig.LDAPServerConfig.UsernameFormat)
+	fmt.Println(usernameDN)
+	// Bind with user credentials to validate the password
+	err = ldapConn.Bind(usernameDN, ldapPassword)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("LDAP authentication failure: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	// Verified user credentials. Now we find the groups they are
+	// a member of.
+	searchSubs := newSubstituter(
+		"username", ldapUsername,
+		"usernamedn", usernameDN,
+	)
+	// We ignore error below as we already validated the search string
+	// at startup.
+	groupSearchFilter, _ := searchSubs.substitute(globalServerConfig.LDAPServerConfig.GroupSearchFilter)
+	searchRequest := ldap.NewSearchRequest(
+		globalServerConfig.LDAPServerConfig.BaseDN,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		groupSearchFilter,
+		[]string{globalServerConfig.LDAPServerConfig.GroupNameAttribute},
+		nil,
+	)
+
+	// fmt.Println("here", groupSearchFilter, globalServerConfig.LDAPServerConfig.BaseDN)
+	sr, err := ldapConn.Search(searchRequest)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("LDAP search failure: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+	groups := []string{}
+	fmt.Println("res:", len(sr.Entries))
+	for _, entry := range sr.Entries {
+		fmt.Printf("Entry DN: %s\n", entry.DN)
+		// We only queried one attribute, so we only look up
+		// the first one.
+		for _, val := range entry.Attributes[0].Values {
+			fmt.Println(val)
+			groups = append(groups, val)
+		}
+	}
+
+	m := map[string]interface{}{
+		// Set expiry time of 1 hour from now.
+		"exp":        UTCNow().Add(1 * time.Hour).Unix(),
```

Comment:
can we take this as ENV? and also add it as part of `config` field? - we can say that this is admin enforced?

Reply:
Yes, will add this part.

Reply:
Done.